### ISSUE TITLE
roles matching in an unexpected way

### ISSUE THREAD

**Issue by gtmtech** [capistrano/capistrano#1381]
Capistrano 3.3.5

I define roles thus:

```
role :myapp_a, some_servers
role :myapp_b, other_servers
role :myapp_all, still_other_servers
```

When specifying in a task:

```
on_roles(:myapp_a)
```

It actually appears to pick up the role for myapp_all. I have tracked this down that the role seems to match based on the fact that the latest role defined starts with "myapp_a", even though it is "myapp_all"

This leads to unexpected behaviour.

**Comment by leehambley** [comment]
Nice catch, I hope you can workaround it by renaming your roles. We'll try and replicate and fix this soon.

**Comment by ayastreb** [comment]
Hi,
I tried to reproduce it with following test:
```ruby
describe 'selecting role by exact name match' do
  before do
    servers.add_role(:role_a, %w{1 2})
    servers.add_role(:role_b, %w{2 3})
    servers.add_role(:role_all, %w{3 4 5})
  end
 
  it 'selects the right roles' do
    expect(servers.roles_for([:role_a]).map(&:hostname)).to eq %w{1 2}
  end
end
```
and it works ok - roles_for([:role_a]) does return servers for role_a, not for role_all.
Could you please provide your exact config when it happens?

**Comment by gtmtech** [comment]
Yep no worries, 

See https://github.com/gtmtech/capistrano3-bugdemo for a complete checkout of the bug.

As you can see in config/deploy/qa.rb

```
require 'bracecomp'

role :myproduct_servers_a, "box{01..02}".expand
role :myproduct_servers_b, "box{03..04}".expand
role :myproduct_servers_all, "box{01..04}".expand
```

the roles should be a=01-02, b=03-04, c=01-04

With task mytask:

```
namespace :mytask do

  desc "bug"
  task :bug do

    puts roles(:myproduct_servers_a)

  end
end
```

I am outputting the value of roles(a) which should be 01-02

However when I run:

```
$ be cap qa mytask:bug
box01
box02
box03
box04
```

I get all 4 boxes.

**Comment by ayastreb** [comment]
Ah, ok, got it.
It seems to be already fixed on master, but still reproduce on 3.3.5
@leehambley you promised to do the release few weeks ago :)

**Comment by leehambley** [comment]
Indeed I did, I really ought to get better at that. It's just so, so difficult to sync that, plus all theplugins and check the docs, etc. I'm making myself a note.

**Comment by ayastreb** [comment]
@leehambley, can I help with anything? I'd be glad to be of use

**Comment by leehambley** [comment]
If you can reproduce in our test suite, and work on a fix that would be amazing, I simply don't have the time, I barely (and often poorly) keep up with the releases and merging the work of all the good people who contrib.

**Comment by ayastreb** [comment]
@leehambley I've already tried to reproduce this issue with a test from my previous comment - it passes on master, but fails on 3.5.5, so I figured this issue was already fixed in master (there was a fix for roles matching).
Would you like me to submit that test as a PR?

**Comment by leehambley** [comment]
Sorry, I missed that bit of history, thanks for your clarification, at least if you could check our testcases (master), it's entirely possible that the change that fixed this introduced a better test case. If you could report back, that'd be great. I'd just like to avoid that fixing this was an accident, and we don't have a test case to cover this circumstance.

**Comment by ayastreb** [comment]
it was fixed in PR #1342 - and that PR has a test for this as well.

**Comment by leehambley** [comment]
Thanks, so I can close this, and try and get a release done today or tomorrow. Watch `@capistranorb` on Twitter.

**Comment by gtmtech** [comment]
Tested and this is fixed in 3.4.0
